revolution_pi/gpio_pin.go

```diff
@@ -11,37 +11,83 @@ import (
 	"unsafe"
 )
 
+const (
+	outputPWMActiveOffset    = 110 // offset for reading a PWM frequency from a DIO
+	outputPWMFrequencyOffset = 112
+	outputWordToPWMOffset    = 2
+	inputWordToCounterOffset = 6
+	dioMemoryOffset          = 88
+)
+
 type gpioPin struct {
-	Name        string // Variable name
-	Address     uint16 // Address of the byte in the process image
-	BitPosition uint8  // 0-7 bit position, >= 8 whole byte
-	Length      uint16 // length of the variable in bits. Possible values are 1, 8, 16 and 32
-	ControlChip *gpioChip
-	pwmMode     bool
-	initialized bool
+	Name         string // Variable name
+	Address      uint16 // Address of the byte in the process image
+	BitPosition  uint8  // 0-7 bit position, >= 8 whole byte
+	Length       uint16 // length of the variable in bits. Possible values are 1, 8, 16 and 32
+	ControlChip  *gpioChip
+	pwmMode      bool
+	initialized  bool
+	outputOffset uint16
+	inputOffset  uint16
 }
 
 func (pin *gpioPin) initialize() error {
-	val, err := pin.ControlChip.getBitValue(int64(110+(pin.Address-70)), pin.BitPosition)
+	dio, err := pin.ControlChip.findDIODevice(pin.Address)
 	if err != nil {
+		pin.ControlChip.logger.Debugf("pin is not from the DIO")
 		return err
 	}
+
+	// if the requested pin is checking the Output WORD. The WORD takes up to 2 bytes
+	pin.outputOffset = dio.i16uOutputOffset
+	pin.inputOffset = dio.i16uInputOffset
+
+	val := false
+
+	// for output gpio pins pwm can be enabled, so we should check for that
+	if pin.isOutputWord() {
+		// if the normal gpio output is given, use the bit position to check if we are in pwm mode
+		val, err = pin.ControlChip.getBitValue(int64(dio.i16uInputOffset+outputPWMActiveOffset+pin.Address%dio.i16uOutputOffset), pin.BitPosition)
+		if err != nil {
+			return err
+		}
+	} else if pin.isOutputPWM() {
+		// we want to read a bit from OutputPWMActive WORD to see if pwm is enabled,
+		// so we convert the pin address into the matching bits, where PWM_1 corresponds to bit 0.
+		// Output pins start at dio.i16uOutputOffset+2, so we can subtract pin address by that amount to get the correct bit
+		pwmActiveBitPosition := uint8(pin.Address - dio.i16uOutputOffset - outputWordToPWMOffset)
+		val, err = pin.ControlChip.getBitValue(int64(dio.i16uInputOffset+outputPWMActiveOffset+uint16(pwmActiveBitPosition/8)), pwmActiveBitPosition%8)
+		if err != nil {
+			return err
+		}
+	}
+
```

Comment:
this is logic to map Output pins with their corresponding pwm pins, to help make them interchangeable when interacting with the rev-pi. removing interchangable logic would simplify this code significantly, but having the logic makes it easier to use for first time users.